In [1]:
#range 를 5로 고정했다.
def solution(answers):
    num1 = [1, 2, 3, 4, 5]
    num2 = [2, 1, 2, 3, 2, 4, 2, 5]
    num3 = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    answer1 = []
    answer2 = []
    answer3 = []
    for i in range(len(answers)):
        if (num1)[i] == (answers)[i]:
            answer1.append(i)
        if (num2)[i] == (answers)[i]:
            answer2.append(i)
        if (num3)[i] == (answers)[i]:
            answer3.append(i)
            if len(answer1) == len(answer2) == len(answer3):
                return [1,2,3]
    return [1]

In [2]:
answers =[1,3,2,4,2]
solution(answers)

[1, 2, 3]

위 풀이는 첫 번째롤 풀었을 때다.ㅜㅜ

In [17]:
%%timeit
def solution1(answers):
    num1 = [1,2,3,4,5]
    num2 = [2,1,2,3,2,4,2,5]
    num3 = [3,3,1,1,2,2,4,4,5,5]
    result = []
    score = [0]*3
    for i in range(len(num1)):
        
        print(f'{i}번째 계산 중입니다')
        # "먼저 2점, 이나 5점을 만들어주는 사람이 1등이다"도 만들어 볼 수 있을 것 같다.
        
        if answers[i] == num1[i]:
            score[0] += 1
            
            print(f'{score}1번 스코어{i}번째') 
            
        if answers[i] == num2[i]:
            score[1] += 1
            
            print(f'{score}2번 스코어{i}번째')
            
        if answers[i] == num3[i]:
            score[2] += 1
            
            print(f'{score}3번 스코어{i}번째')
    print(f'answer{answers}에 맞는\n1번,2번,3번 학생의 점수는\n{score} 입니다')
    
    for i, s in enumerate(score):
        if s == max(score):
            result.append(i+1)
            
            print(f'result가{result}{i+1}번째 계산만에 나왔다')
            
    return result

68.2 ns ± 0.199 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [18]:
answers =[1,3,2,4,2]
solution1(answers)

0번째 계산 중입니다
[1, 0, 0]1번 스코어0번째
1번째 계산 중입니다
[1, 0, 1]3번 스코어1번째
2번째 계산 중입니다
[1, 1, 1]2번 스코어2번째
3번째 계산 중입니다
[2, 1, 1]1번 스코어3번째
4번째 계산 중입니다
[2, 2, 1]2번 스코어4번째
[2, 2, 2]3번 스코어4번째
answer[1, 3, 2, 4, 2]에 맞는
1번,2번,3번 학생의 점수는
[2, 2, 2] 입니다
result가[1]1번째 계산만에 나왔다
result가[1, 2]2번째 계산만에 나왔다
result가[1, 2, 3]3번째 계산만에 나왔다


[1, 2, 3]

In [5]:
# 다른사람들의 풀이 연구

In [26]:
%%timeit
def solution2(answers):
    pattern1 = [1,2,3,4,5]
    pattern2 = [2,1,2,3,2,4,2,5]
    pattern3 = [3,3,1,1,2,2,4,4,5,5]
    score = [0, 0, 0]
    result = []

    for idx, answer in enumerate(answers):
        if answer == pattern1[idx%len(pattern1)]:
            print(f'answer->{answer} pattern1->{pattern1} idx->{idx} lenpattern1->{len(pattern1)}')
            score[0] += 1
        if answer == pattern2[idx%len(pattern2)]:
            print(f'answer->{answer} pattern2->{pattern2} idx->{idx} lenpattern2->{len(pattern2)}')
            score[1] += 1
        if answer == pattern3[idx%len(pattern3)]:
            print(f'answer->{answer} pattern3->{pattern3} idx->{idx} lenpattern3->{len(pattern3)}')
            score[2] += 1

    for idx, s in enumerate(score):
        if s == max(score):
            result.append(idx+1)
    
    return result

69.1 ns ± 0.913 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [27]:
answers =[1,3,2,4,2]
solution2(answers)

answer->1 pattern1->[1, 2, 3, 4, 5] idx->0 lenpattern1->5
answer->3 pattern3->[3, 3, 1, 1, 2, 2, 4, 4, 5, 5] idx->1 lenpattern3->10
answer->2 pattern2->[2, 1, 2, 3, 2, 4, 2, 5] idx->2 lenpattern2->8
answer->4 pattern1->[1, 2, 3, 4, 5] idx->3 lenpattern1->5
answer->2 pattern2->[2, 1, 2, 3, 2, 4, 2, 5] idx->4 lenpattern2->8
answer->2 pattern3->[3, 3, 1, 1, 2, 2, 4, 4, 5, 5] idx->4 lenpattern3->10


[1, 2, 3]

위에서 나와 다른점은 검증하는 곳인 것 같다. 나는 결과를 알고 있기에 5번만 반복하여 결과를 내었고  
위 분은 answer과 pattern을 모두 비교해 검증을 한 것이다.  
결과를 다가가기 위해서 주어진 문제인식과 조건들을 좀 더 신중히 살펴봐야겠다.  

In [25]:
%%timeit
def solution(answers):
    p = [[1, 2, 3, 4, 5],
         [2, 1, 2, 3, 2, 4, 2, 5],
         [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]]
    s = [0] * len(p)

    for q, a in enumerate(answers):
        for i, v in enumerate(p):
            if a == v[q % len(v)]:
                s[i] += 1
    return [i + 1 for i, v in enumerate(s) if v == max(s)]


68.6 ns ± 0.228 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


결국 같은 형식으로 풀이 된 것 이다. 단순해보여서 똑똑해보이지만 우리들이 봤을때는     
시각적으로 봤을 때 이분이 어떻게 풀었는지 판단이 바로 안된다.  

me = 68.2 ns ± 0.199 ns  
other1 = 69.1 ns ± 0.913 ns   
other2 = 68.6 ns ± 0.228 ns   

하지만 컴퓨터가 계산할때는 3번째 분이 제일 빠르다고 나왔다.  
우리가 볼때는 1번이 보기좋다  
컴퓨터가 볼땐 2번이 보기좋다   
